src/schemes/boundary/open_boundary/system.jl

```diff
@@ -0,0 +1,622 @@
+    @nospecialize system # reduce precompilation time
+
+    if get(io, :compact, false)
+        show(io, system)
+    else
+        summary_header(io, "OpenBoundarySPHSystem{$(ndims(system))}")
+        if system.buffer isa SystemBuffer
+            summary_line(io, "#particles", nparticles(system))
+            summary_line(io, "#buffer_particles", system.buffer.buffer_size)
+        else
+            summary_line(io, "#particles", nparticles(system))
+        end
+        summary_line(io, "boundary", system.boundary_zone)
+        summary_line(io, "flow direction", system.flow_direction)
+        summary_line(io, "prescribed velocity", string(nameof(system.reference_velocity)))
+        summary_line(io, "prescribed pressure", string(nameof(system.reference_pressure)))
+        summary_line(io, "prescribed density", string(nameof(system.reference_density)))
+        summary_line(io, "width", round(norm(system.spanning_set[1]), digits=3))
+        summary_footer(io)
+    end
+end
+
+callback_used!(system::OpenBoundarySPHSystem) = system.callback_used[] = true
+
+@inline source_terms(system::OpenBoundarySPHSystem) = nothing
+
+@inline hydrodynamic_mass(system::OpenBoundarySPHSystem, particle) = system.mass[particle]
+
+@inline function particle_density(v, system::OpenBoundarySPHSystem, particle)
+    return system.density[particle]
+end
+
+@inline function particle_pressure(v, system::OpenBoundarySPHSystem, particle)
+    return system.pressure[particle]
+end
+
+function spanning_vectors(plane_points, zone_width)
+
+    # Convert to tuple
+    return spanning_vectors(tuple(plane_points...), zone_width)
+end
+
+function spanning_vectors(plane_points::NTuple{2}, zone_width)
+    plane_size = plane_points[2] - plane_points[1]
+
+    # Calculate normal vector of plane
+    b = Vector(normalize([-plane_size[2]; plane_size[1]]) * zone_width)
+
+    return hcat(b, plane_size)
+end
+
+function spanning_vectors(plane_points::NTuple{3}, zone_width)
+    # Vectors spanning the plane
+    edge1 = plane_points[2] - plane_points[1]
+    edge2 = plane_points[3] - plane_points[1]
+
+    if !isapprox(dot(edge1, edge2), 0.0, atol=1e-7)
+        throw(ArgumentError("the provided points do not span a rectangular plane"))
+    end
+
+    # Calculate normal vector of plane
+    c = Vector(normalize(cross(edge2, edge1)) * zone_width)
+
+    return hcat(c, edge1, edge2)
+end
+
+@inline function within_boundary_zone(particle_coords, system)
+    (; zone_origin, spanning_set) = system
+    particle_position = particle_coords - zone_origin
+
+    for dim in 1:ndims(system)
+        span_dim = spanning_set[dim]
+        # Checks whether the projection of the particle position
+        # falls within the range of the zone.
+        if !(0 <= dot(particle_position, span_dim) <= dot(span_dim, span_dim))
+
+            # Particle is not in boundary zone.
+            return false
+        end
+    end
+
+    # Particle is in boundary zone.
+    return true
+end
```

Comment:
FYI: I can make this faster by adding a flag that the boundary zone is axis-aligned and only check the coordinates. However, Idk if this is necessary?